art/attacks/evasion/pixel_threshold.py

```diff
@@ -0,0 +1,1346 @@
+        Generate adversarial samples and return them in an array.
+
+        :param x: An array with the original inputs.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels) one-hot-encoded of shape
+                  (nb_samples, nb_classes) or indices of shape (nb_samples,).
+                  Only provide this parameter if you'd like to use true labels
+                  when crafting adversarial samples. Otherwise, model
+                  predictions are used as labels to avoid the "label leaking"
+                  effect
+                  (explained in this paper: https://arxiv.org/abs/1611.01236).
+                  Default is `None`.
+        :type y: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        y = check_and_transform_label_format(
+            y, self.classifier.nb_classes(), return_one_hot=False)
+
+        if y is None:
+            if self.targeted:
+                raise ValueError(
+                    'Target labels `y` need to be provided for a targeted '
+                    'attack.'
+                )
+            y = np.argmax(self.classifier.predict(x), axis=1)
+        else:
+            if len(y.shape) > 1:
+                y = np.argmax(y, axis=1)
+
+        if self.th is None:
+            logger.info('Performing minimal perturbation Attack.')
+
+        if isinstance(self.classifier, PyTorchClassifier):
+            x = np.swapaxes(x, 1, 3).astype(np.float32)
+
+        if np.max(x) <= 1:
+            x = x * 255.
+
+        adv_x_best = []
+        for image, target_class in zip(x, y):
+            if self.th is None:
+                self.min_th = 127
+                start, end = 1, 127
+                while True:
+                    image_result = []
+                    threshold = (start + end) // 2
+                    success, trial_image_result = self.attack(
+                        image, target_class, threshold)
+                    if image_result or success:
+                        image_result = trial_image_result
+                    if success:
+                        end = threshold - 1
+                    else:
+                        start = threshold + 1
+                    if success:
+                        self.min_th = threshold
+                    if end < start:
+                        break
+            else:
+                success, image_result = self.attack(
+                    image, target_class, self.th)
+            adv_x_best += [image_result]
+
+        adv_x_best = np.array(adv_x_best)
+
+        if np.max(x) <= 1:
+            x = x / 255.
+
+        if isinstance(self.classifier, PyTorchClassifier):
+            x = np.swapaxes(x, 1, 3).astype(np.float32)
+            adv_x_best = np.swapaxes(adv_x_best, 1, 3).astype(np.float32)
+
+        if y is not None:
+            y = to_categorical(y, self.classifier.nb_classes())
+
+        rate_best = 100 * \
```

Comment:
Please integrate the calculation of `rate_best` into `logger.info` in line 193, that way it is only calculated if the logging information is actually logged.

Reply:
Integrated.